woocommerce/woocommerce-ios
# [Subscriptions] Refactor localized descriptions for product subscription settings

Closes: #9394


## Description

This is a followup to @Ecarrion's advice in https://github.com/woocommerce/woocommerce-ios/pull/9486#pullrequestreview-1390620935, to avoid creating extensions on networking entities for content that is needed in specific views. It removes the `ProductSubscription` extension and instead moves these descriptions to the specific views that need them.

### Changes

* The file `ProductSubscription+UI` was replaced with `SubscriptionPeriod+UI`. This retains the `SubscriptionPeriod` extension (which contains only localized descriptions for each case in that `enum`) and removes the `ProductSubscription` extension.
* The strings used in the Subscription row in product/variation details are now part of `DefaultProductFormTableViewModel`:
   * This also makes a change to the details in the Subscription row if a variation has no price set. Instead of showing `Regular price: No price set` the row omits the price details, showing only the expiry details. This is consistent with how we show product/variation details in other rows (hiding details that aren't set).
* The strings used in the Subscription settings view are now part of `SubscriptionSettingsViewModel`:
   * This view model now also has a convenience initializer that takes a `ProductSubscription` and creates the appropriate description strings based on the subscription properties.
* The unit tests are moved to the relevant test classes (`DefaultProductFormTableViewModelTests` and `SubscriptionSettingsViewModelTests`).

## Testing instructions


### Prerequisites

1. Install and activate the [Subscriptions extension](https://woocommerce.com/products/woocommerce-subscriptions/) on your store.
2. Create at least one product with the Variable Subscription product type.
3. Add at least one variation to the product.

### To test

1. Build and run the app.
2. Go to the Products tab.
4. Open a variable subscription product and select Variations.
5. Select a variation.
6. Confirm the Subscription row appears with the expected variation details.
7. Select Subscription and confirm the Subscription details appear with the expected subscription settings.

You can also repeat these steps with a simple subscription product, and confirm the expected details appear in the Subscription row and settings view.

## Screenshots

**Variable Subscription**
Variation Details|Subscription Settings
-|-
![Simulator Screen Shot - iPhone 14 Pro - 2023-04-19 at 10 06 42](https://user-images.githubusercontent.com/8658164/233026417-0fcdc4d5-3d51-4349-841d-14d4cc3c1ffd.png)|![Simulator Screen Shot - iPhone 14 Pro - 2023-04-19 at 10 06 44](https://user-images.githubusercontent.com/8658164/233026433-1cd12427-5a59-4d73-b9a5-9cab1108c8fb.png)
![Simulator Screen Shot - iPhone 14 Pro - 2023-04-19 at 10 06 33](https://user-images.githubusercontent.com/8658164/233026456-f3f48f07-d258-4cfb-a8d3-933272b81872.png)|![Simulator Screen Shot - iPhone 14 Pro - 2023-04-19 at 10 06 36](https://user-images.githubusercontent.com/8658164/233026475-f464e346-544b-4907-b2e3-a89a13f5eb4e.png)

**Simple Subscription**
Product Details|Subscription Settings
-|-
![Simulator Screen Shot - iPhone 14 Pro - 2023-04-19 at 10 06 20](https://user-images.githubusercontent.com/8658164/233026379-2f6a5c32-e8c6-41dc-a372-22d683186131.png)|![Simulator Screen Shot - iPhone 14 Pro - 2023-04-19 at 10 06 23](https://user-images.githubusercontent.com/8658164/233026401-14f283f5-c5a2-4dae-b0cc-2c36ea035556.png)
---
